File: src/141.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

'''
基本思路，边遍历边保存
如果碰见之间保存过的节点，就有环
time defeat: 5.12%
space defeat: 84.38%
time consuming: less than 20 mins
'''
class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        appearNode = list()
        #极端情况
        if not head or head.next == None:
            return False
        else:
            #首节点入队
            appearNode.append(head)
            nowNode = head.next
            #逐节点遍历，直至最后一个
            while nowNode != None:
                if nowNode in appearNode:
                    #出现环
                    return True
                else:
                    appearNode.append(nowNode)
                    nowNode = nowNode.next
            return False
        

File: src/141.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        //干翻字节！
        //应该用双指针，但是要前进几步我给忘了，双指针的空间复杂度就是O1了
        //用相遇问题吧
        //对了，自己做的，就是用相遇的想法
        //十分钟，两次提交，时间击败-99.79%，空间击败-67.13%
        //边界处理
        if(head == nullptr || head->next == nullptr){
            return false;
        }
        //快慢双指针
        ListNode * slow = head;
        ListNode * fast = head->next;
        //开始冲锋
        while(fast&& fast->next){
            if(fast == slow){
                //相遇
                return true;
            }
            slow = slow->next;
            fast = fast->next->next;
        }
        return false;
    }
};
